File: python/045.py
#!/usr/bin/env python3
"""
Project Euler Problem 45: Triangular, pentagonal, and hexagonal

Find the next triangle number after 40755 that is also pentagonal and hexagonal.
"""

from math import isqrt


def triangular(n: int) -> int:
    return n * (n + 1) // 2


def is_pentagonal(n: int) -> bool:
    d = 1 + 24 * n
    root = isqrt(d)
    return root * root == d and (1 + root) % 6 == 0


def is_hexagonal(n: int) -> bool:
    d = 1 + 8 * n
    root = isqrt(d)
    return root * root == d and (1 + root) % 4 == 0


def main():
    n = 286  # Starting after T_285
    while True:
        t = triangular(n)
        if is_pentagonal(t) and is_hexagonal(t):
            print(t)
            break
        n += 1


if __name__ == "__main__":
    main()

File: c/045.c
#include <stdio.h>
#include <stdbool.h>
#include <math.h>

bool is_pentagonal(long long num) {
    if (num <= 0) return false;
    long long disc = 1 + 24 * num;
    long long candidate = (long long)round(sqrt((double)disc));
    if (candidate * candidate != disc) return false;
    return (1 + candidate) % 6 == 0;
}

bool is_hexagonal(long long num) {
    if (num <= 0) return false;
    long long disc = 1 + 8 * num;
    long long candidate = (long long)round(sqrt((double)disc));
    if (candidate * candidate != disc) return false;
    return (1 + candidate) % 4 == 0;
}

int main(void) {
    long long known = 40755;
    int n = 286;

    while (1) {
        long long triangle = (long long)n * (n + 1) / 2;
        if (triangle > known && is_pentagonal(triangle) && is_hexagonal(triangle)) {
            printf("%lld\n", triangle);
            break;
        }
        n++;
    }

    return 0;
}
